server/src/main/java/org/elasticsearch/action/ActionModule.java

```diff
@@ -825,6 +826,9 @@ public void initRestHandlers(Supplier<DiscoveryNodes> nodesInCluster, Predicate<
         Predicate<AbstractCatAction> catActionsFilter = restExtension.getCatActionsFilter();
         Predicate<RestHandler> restFilter = restExtension.getActionsFilter();
         Consumer<RestHandler> registerHandler = handler -> {
+            if (handler instanceof BaseRestHandler baseRestHandler && threadPool != null) {
+                baseRestHandler.setThreadContext(threadPool.getThreadContext());
```

Comment:
I don't think we should add a mutable member. We've strived to keep things like rest handlers immutable. They should be stateless.

The way we've handled common things to utilize within the rest handlers before is to  pass them in to handleRequest. This is why the NodeClient is passed there. Could that approach be used? If so, we should do that as a separate change because it will affect the signature of all rest handlers.

Reply:
What do you mean by "transport headers"? Headers are part of http requests, transport protocol doesn't have any such concept. If you mean using thread context to pass additional information to transport requests, why wouldn't we adjust the transport requests? These are internal and can change freely.

Reply:
The idea of using the context was motivated by us having a bunch of spots where we eventually want to fix the same thing. The second most important one after this is errors in bulk responses.
Since we already have this as a generic header on all REST requests we figured it might be ok to use the thread context to propagate this information generically as well.
It's also just a lot easier to do it this way for us just from the mechanics of our code. We have so many different transport APIs, it would require some effort fixing all of them :)

Reply:
Keep me honest here, but in OutboundMessage we have the writeVariableHeader method, that serialises `ThreadContext.defaultHeader` to the wire, so that every transport message has, besides the request "body" also headers from the threadpool context. 
I agree another possible solution would be to not use the header but instead change the transport request but I wanted to build a more general approach so that we can use the same mechanism in the future for other endpoints.
I think this should not be a transport request change, since both request and response can be the exact same. We just need to inform the datanode that we don't want stack trace and that (IMO) can be done via headers.

Reply:
> The way we've handled common things to utilize within the rest handlers before is to pass them in to handleRequest. This is why the NodeClient is passed there. Could that approach be used? If so, we should do that as a separate change because it will affect the signature of all rest handlers.

Actually this is easier than expected, thanks for the input, we have threadpool directly in the NodeClient so I can simply use that 😄

Reply:
@rjernst I think I addressed the main concern you had about rest handlers immutability and since the implementation does not need signature changes for all the handlers I think we can avoid opening a different PR.
One question I have is if you are okay with the usage we did of ThreadContext.